src/Application/Matching/Allocator.php

```diff
@@ -0,0 +1,229 @@
+<?php
+
+declare(strict_types=1);
+
+namespace MatchBot\Application\Matching;
+
+use Doctrine\DBAL\Exception as DBALException;
+use Doctrine\ORM\EntityManagerInterface;
+use MatchBot\Application\Assertion;
+use MatchBot\Domain\CampaignFunding;
+use MatchBot\Domain\CampaignFundingRepository;
+use MatchBot\Domain\Donation;
+use MatchBot\Domain\FundingWithdrawal;
+use Psr\Log\LoggerInterface;
+
+class Allocator
+{
+    /**
+     * Seems we must do this dynamically to avoid circular dependency. Remains null for some tests for now.
+     */
+    private ?CampaignFundingRepository $campaignFundingRepository = null;
+
+    public function __construct(
+        private Adapter $adapter,
+        private EntityManagerInterface $entityManager,
+        private LoggerInterface $logger,
+    ) {
+        // For now, just type this as a few tests don't set it up. We assert before use in allocateMatchFunds() where
+        // it's really needed.
+        /** @var CampaignFundingRepository $campaignFundingRepo */ // @phpstan-ignore varTag.type
```

Comment:
Not sure why phpstan doesn't like this type. Adding the `<>` makes it complain about too many template parts..

Reply:
phpstan wants us to do `\assert($campaignFundingRepo instanceof CampaignFundingRepository);` on the next line to make sure the type is what we say it is instead of just using the var tag to suggest it. I'm not sure where you were adding `<>`, CampaignFundingRepository is not generic.

Reply:
Yes, that's what I wanted to do but it would have involved more test work as the comment says. Hopefully not needed in any case if we can successfully change the constructor DI.